shared/chat/conversation/index-hoc.js

```diff
@@ -65,13 +88,65 @@ const hoc = withProps(
       isLoading,
       onAttach,
       onEditMessage,
-      onEditLastMessage: () => console.log('todo'),
+      onEditLastMessage,
+      onUnmountText: onStoreInputText,
+      focusInputCounter: focusInputCounter,
       onPostMessage,
       selectedConversation,
     }
 
-    return {inputProps, listProps}
+    const headerProps: HeaderProps = {
+      muted,
+      onBack,
+      onOpenFolder,
+      onShowProfile,
+      onToggleSidePanel,
+      sidePanelOpen,
+      users,
+    }
+
+    return {inputProps, listProps, headerProps, onOpenNewerConversation}
   }
 )
 
+function _decorateSupersedes (props: Props, messages: Immutable.List<Constants.Message>): Immutable.List<Constants.Message> {
+  if (props.supersedes && !props.moreToLoad) {
+    const {conversationIDKey, finalizeInfo: {resetUser}} = props.supersedes
+    const supersedesMessage: Constants.SupersedesMessage = {
+      type: 'Supersedes',
+      supersedes: conversationIDKey,
+      username: resetUser,
+      timestamp: Date.now(),
+      key: `supersedes-${conversationIDKey}-${resetUser}`,
+    }
+    return messages.unshift(supersedesMessage)
+  }
+
+  return messages
+}
+
+const decoratedMesssagesHoc = withProps((props) => ({
+  messages: _decorateSupersedes(props, props.messages),
+}))
+
+const focusInputHoc = compose(
+  withState(
+    'focusInputCounter',
+    'setFocusInputCounter',
+    0
+  ),
+  withProps(({setFocusInputCounter}) => ({onFocusInput: () => setFocusInputCounter(n => n + 1)})),
+)
+
+const editLastMessageHoc = compose(
+  withState(
+    'editLastMessageCounter',
+    'setEditLastMessageCounter',
+    0
+  ),
+  withProps(({setEditLastMessageCounter}) => ({onEditLastMessage: () => setEditLastMessageCounter(n => n + 1)}))
```

Comment:
This is rather subjective: personally I think putting effects in `componentDidUpdate` is worse than the annoyances of calling methods on refs. We should discuss this more if this pattern becomes prevalent.